src/main/java/edu/uwaterloo/ece658/service/DataHandlerService.java

```
@@ -0,0 +1,160 @@
+            String username, 
+            String S3key,
+            String Md5,
+            Integer photoSize, 
+            Date uploadDate,
+            boolean isPrivatePhoto,
+            List<String> tagNames)
+    {
+        // store key in table with tags
+        List<Tag> tags = initializeTags(tagNames);
+        User user = userFacade.findUserByUserName(username);
+        Photo photo = initializePhoto(
+                        S3key,
+                        Md5,
+                        photoSize,
+                        uploadDate,
+                        isPrivatePhoto);
+        
+        photo.setTags(tags);
+        photo.addUser(user);
+        
+        user.addUploadedPhoto(photo);
+        userFacade.edit(user);
+        
+        for (Tag tag : tags) {
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);
+        }
+        
+        // call notify to send messages to users
+        return subscriptionService.notifyUsers(tags, photo);
+    }
+    
+    public void removeTagFromPhoto(String username, String S3key, String tagName) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        Tag tag = tagFacade.getNormalTagByName(tagName);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            photo.getTags().remove(tag);
+            tag.getPhotosUnderThisTag().remove(photo);
+            tagFacade.edit(tag);
+            photoFacade.edit(photo);
+        }
+    }
+    
+    public void addTagToPhoto(String username, String S3key, String tagName) {
+        Photo photo = photoFacade.getPhotoByS3Key(S3key);
+        User user = userFacade.findUserByUserName(username);
+        Tag tag = initializeTag(tagName);
+        
+        if (photo.getUploadedUsers().contains(user)) {
+            photo.addTag(tag);
+            tag.addPhoto(photo);
+            tagFacade.edit(tag);
+            photoFacade.edit(photo);
+        }

```

Can you add a TODO for else case to say instead of silently ignoring, throw business logic exception?

---

yup